test/jdk/java/lang/ProcessBuilder/JspawnhelperProtocol.java

```diff
@@ -0,0 +1,176 @@
+ */
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import jdk.test.lib.process.ProcessTools;
+
+public class JspawnhelperProtocol {
+    // Timout in seconds
+    private static final int TIMEOUT = 60;
+    private static final int ERROR = 10;
+    private static final String[] CMD = { "pwd" };
+    private static final String ENV_KEY = "JTREG_JSPAWNHELPER_PROTOCOL_TEST";
+
+    private static void childCode(String arg) throws IOException, InterruptedException {
+        System.out.println("Recursively executing 'JspawnhelperProtocol " + arg + "'");
+        Process p = null;
+        try {
+            p = Runtime.getRuntime().exec(CMD);
+        } catch (Exception e) {
+            System.out.println(e.toString());
+            System.exit(ERROR);
+        }
+        if (!p.waitFor(TIMEOUT, TimeUnit.SECONDS)) {
+            System.exit(ERROR + 1);
+        }
+        if (p.exitValue() == 0) {
+            String pwd = p.inputReader().readLine();
+            if (!Path.of("").toAbsolutePath().toString().equals(pwd)) {
+                System.exit(ERROR + 2);
+            }
+            System.out.println("  Successfully executed '" + CMD[0] + "'");
+            System.exit(0);
+        } else {
+            System.out.println("  Failed to executed '" + CMD[0] + "' (exitValue=" + p.exitValue() + ")");
+            System.exit(ERROR + 3);
+        }
+    }
+
+    private static void normalExec() throws Exception {
+        ProcessBuilder pb;
+        pb = ProcessTools.createJavaProcessBuilder("-Djdk.lang.Process.launchMechanism=posix_spawn",
+                                                   "JspawnhelperProtocol",
+                                                   "normalExec");
+        Process p = pb.start();
+        if (!p.waitFor(TIMEOUT, TimeUnit.SECONDS)) {
+            System.exit(ERROR + 4);
+        }
+        if (p.exitValue() != 0) {
+            throw new Exception("Child exited with " + p.exitValue());
+        }
+        System.out.println(new String(p.getInputStream().readAllBytes()));
+    }
+
+    private static void simulateCrashInChild(int stage) throws Exception {
+        ProcessBuilder pb;
+        pb = ProcessTools.createJavaProcessBuilder("-Djdk.lang.Process.launchMechanism=posix_spawn",
+                                                   "JspawnhelperProtocol",
+                                                   "simulateCrashInChild" + stage);
+        pb.environment().put(ENV_KEY, Integer.toString(stage));
+        Process p = pb.start();
+
+        BufferedReader br = p.inputReader();
+        String line = br.readLine();
+        boolean foundCrashInfo = false;
+        while (line != null) {
+            System.out.println(line);
+            if (line.equals("posix_spawn:0")) {
+                foundCrashInfo = true;
+            }
+            line = br.readLine();
+        }
```

Comment:
Try-with-resources works well in cases like this.  (Moving foundCrashInfo out of the t-w-r).

Reply:
Done.